File: update-core/src/main/java/dev/bee/kanjianki/updatecore/PackageInstallStatusPolicy.java
package dev.bee.kanjianki.updatecore;

public final class PackageInstallStatusPolicy {
    public static final int STATUS_SUCCESS = 0;
    public static final int STATUS_PENDING_USER_ACTION = -1;
    public static final int ANDROID_S_API_LEVEL = 31;
    public static final int ANDROID_T_API_LEVEL = 33;
    public static final int ANDROID_U_API_LEVEL = 34;
    public static final int ANDROID_V_API_LEVEL = 35;
    public static final int ANDROID_B_API_LEVEL = 36;
    public static final String SOURCE_MANUAL = "MANUAL";
    public static final String SOURCE_AUTOMATIC = "AUTOMATIC";
    public static final String SOURCE_CACHED = "CACHED";

    private PackageInstallStatusPolicy() {
    }

    public static InstallCallback mapInstallStatus(int status, String message) {
        if (status == STATUS_SUCCESS) {
            return new InstallCallback(false, true, "Install finished.");
        }
        if (status == STATUS_PENDING_USER_ACTION) {
            return new InstallCallback(true, false, "Android needs confirmation to finish installing.");
        }
        String suffix = message == null || message.trim().isEmpty() ? "" : ": " + message.trim();
        return new InstallCallback(false, false, "Install failed" + suffix + ".");
    }

    public static String sourceNameOrDefault(String raw) {
        if (SOURCE_MANUAL.equals(raw) || SOURCE_AUTOMATIC.equals(raw) || SOURCE_CACHED.equals(raw)) {
            return raw;
        }
        return SOURCE_AUTOMATIC;
    }

    public static boolean shouldLaunchInstallConfirmation(String sourceName) {
        String normalized = sourceNameOrDefault(sourceName);
        return SOURCE_MANUAL.equals(normalized) || SOURCE_CACHED.equals(normalized);
    }

    public static boolean shouldAllowInstallerWithoutExtraUserAction(int targetSdk, int runtimeSdk) {
        return targetSdk >= minimumTargetSdkForInstallerWithoutExtraUserAction(runtimeSdk);
    }

    public static int minimumTargetSdkForInstallerWithoutExtraUserAction(int runtimeSdk) {
        if (runtimeSdk >= ANDROID_B_API_LEVEL) {
            return ANDROID_U_API_LEVEL;
        }
        if (runtimeSdk >= ANDROID_V_API_LEVEL) {
            return ANDROID_T_API_LEVEL;
        }
        if (runtimeSdk >= ANDROID_U_API_LEVEL) {
            return ANDROID_S_API_LEVEL;
        }
        if (runtimeSdk >= ANDROID_T_API_LEVEL) {
            return 30;
        }
        if (runtimeSdk >= ANDROID_S_API_LEVEL) {
            return 29;
        }
        return Integer.MAX_VALUE;
    }

    public static final class InstallCallback {
        public final boolean pendingUserAction;
        public final boolean success;
        public final String message;

        private InstallCallback(boolean pendingUserAction, boolean success, String message) {
            this.pendingUserAction = pendingUserAction;
            this.success = success;
            this.message = message;
        }

        public boolean pendingUserAction() {
            return pendingUserAction;
        }

        public boolean success() {
            return success;
        }

        public String message() {
            return message;
        }
    }
}

File: update-core/src/main/kotlin/dev/bee/kanjianki/updatecore/PackageInstallStatusPolicy.kt
package dev.bee.kanjianki.updatecore

import java.util.Locale

object PackageInstallStatusPolicy {
    const val STATUS_SUCCESS = 0
    const val STATUS_PENDING_USER_ACTION = -1
    const val ANDROID_S_API_LEVEL = 31
    const val ANDROID_T_API_LEVEL = 33
    const val ANDROID_U_API_LEVEL = 34
    const val ANDROID_V_API_LEVEL = 35
    const val ANDROID_B_API_LEVEL = 36
    const val SOURCE_MANUAL = "MANUAL"
    const val SOURCE_AUTOMATIC = "AUTOMATIC"
    const val SOURCE_CACHED = "CACHED"

    private const val JAPANESE_LANGUAGE = "ja"

    @JvmStatic
    fun mapInstallStatus(status: Int, message: String?): InstallCallback {
        if (status == STATUS_SUCCESS) {
            return InstallCallback(false, true, installFinishedMessage())
        }
        if (status == STATUS_PENDING_USER_ACTION) {
            return InstallCallback(true, false, installPermissionNeededMessage())
        }
        return InstallCallback(false, false, installFailedMessage(message))
    }

    @JvmStatic
    fun installFinishedMessage(): String = localizedText("Install finished.", "インストールが完了しました。")

    @JvmStatic
    fun installPermissionNeededMessage(): String = localizedText(
        "Android needs permission to finish installing.",
        "インストールを完了するにはAndroidの許可が必要です。",
    )

    @JvmStatic
    fun installFailedMessage(message: String?): String {
        val suffix = if (message.isNullOrBlank()) "" else ": ${message.trim()}"
        return localizedText("Install failed$suffix.", "インストールに失敗しました$suffix。")
    }

    @JvmStatic
    fun sourceNameOrDefault(raw: String?): String {
        if (raw == SOURCE_MANUAL || raw == SOURCE_AUTOMATIC || raw == SOURCE_CACHED) {
            return raw
        }
        return SOURCE_AUTOMATIC
    }

    @JvmStatic
    fun shouldLaunchInstallConfirmation(sourceName: String?): Boolean {
        val normalized = sourceNameOrDefault(sourceName)
        return normalized == SOURCE_MANUAL || normalized == SOURCE_CACHED
    }

    @JvmStatic
    fun shouldAllowInstallerWithoutExtraUserAction(targetSdk: Int, runtimeSdk: Int): Boolean {
        return targetSdk >= minimumTargetSdkForInstallerWithoutExtraUserAction(runtimeSdk)
    }

    @JvmStatic
    fun minimumTargetSdkForInstallerWithoutExtraUserAction(runtimeSdk: Int): Int {
        if (runtimeSdk >= ANDROID_B_API_LEVEL) {
            return ANDROID_U_API_LEVEL
        }
        if (runtimeSdk >= ANDROID_V_API_LEVEL) {
            return ANDROID_T_API_LEVEL
        }
        if (runtimeSdk >= ANDROID_U_API_LEVEL) {
            return ANDROID_S_API_LEVEL
        }
        if (runtimeSdk >= ANDROID_T_API_LEVEL) {
            return 30
        }
        if (runtimeSdk >= ANDROID_S_API_LEVEL) {
            return 29
        }
        return Int.MAX_VALUE
    }

    private fun localizedText(english: String, japanese: String): String {
        return if (Locale.getDefault().language == JAPANESE_LANGUAGE) japanese else english
    }

    class InstallCallback(
        @JvmField val pendingUserAction: Boolean,
        @JvmField val success: Boolean,
        @JvmField val message: String,
    ) {
        fun pendingUserAction(): Boolean = pendingUserAction
        fun success(): Boolean = success
        fun message(): String = message
    }
}
